lib/src/license_detection/license.dart

```diff
@@ -0,0 +1,261 @@
+  final String identifier;
+
+  /// Original text from the license file.
+  final String content;
+
+  /// Normalized [Token]s created from the original text.
+  final List<Token> tokens;
+
+  /// Map from [Token.value] to their number of occurrences in this license.
+  final Map<String, int> occurrences;
+
+  License._(this.content, this.tokens, this.occurrences, this.identifier);
+
+  factory License.parse(String identifier, String content) {
+    final tokens = tokenize(content);
+    final table = generateFrequencyTable(tokens);
+    return License._(content, tokens, table, identifier);
+  }
+}
+
+String tokensNormalizedValue(Iterable<Token> tokens) {
+  return tokens.map((token) => token.value).join(' ');
+}
+
+@sealed
+class NGram {
+  /// Text for which the hash value was generated.
+  final String text;
+
+  /// [CRC-32][1] checksum value generated for text.
+  ///
+  /// [1]: https://en.wikipedia.org/wiki/Cyclic_redundancy_check
+  final int checksum;
+
+  /// Index of the first token in the checksum.
+  final int start;
+
+  /// Index of the last token in the checksum.
+  final int end;
+
+  int get granularity => text.split(' ').length;
+
+  NGram(this.text, this.checksum, this.start, this.end);
+}
+
+/// A [License] instance with generated [nGrams].
+// TODO: Change class name to something more meaningful.
+@sealed
+class LicenseWithNGrams extends License {
+  final List<NGram> nGrams;
+
+  final Map<int, List<NGram>> checksumMap;
+
+  final int n;
+
+  LicenseWithNGrams._(
+    this.nGrams,
+    this.checksumMap,
+    this.n,
+    String identifier,
+    String content,
+    List<Token> tokens,
+    Map<String, int> table,
+  ) : super._(
+          content,
+          tokens,
+          table,
+          identifier,
+        );
+
+  factory LicenseWithNGrams.parse(License license, int n) {
+    final nGrams = generateChecksums(license.tokens, n);
+    final table = generateChecksumMap(nGrams);
+    return LicenseWithNGrams._(
+      nGrams,
+      table,
+      license.tokens.length < n ? license.tokens.length : n,
+      license.identifier,
+      license.content,
+      license.tokens,
+      license.occurrences,
+    );
+  }
+}
+
+/// Contains details regarding the results of corpus license match with unknown text.
+@sealed
+class LicenseMatch {
+  /// Sequence of tokens from input text that were considered a match for the detected [License].
+  final List<Token> tokens;
+
+  /// [Diff]s calculated between target tokens and [license] tokens.
+  final List<Diff> diffs;
+
+  /// Range of [diffs] which represents the diffs between known license and unknown license.
+  ///
+  /// Diffs lying outside of [diffRange] represent the text in unknown license
+  /// that is not a part of the [license](source) text.
+  final Range diffRange;
+
+  /// Confidence score of the detected license.
+  final double confidence;
+
+  /// Detected license the input have been found to match with given [confidence].
+  final LicenseWithNGrams license;
```

Comment:
Couldn't this just be `License`, do we need to expose `LicenseWithNGrams` publicly?